datafusion/src/sql/utils.rs

```diff
@@ -390,6 +390,42 @@ pub(crate) fn extract_aliases(exprs: &[Expr]) -> HashMap<String, Expr> {
         .collect::<HashMap<String, Expr>>()
 }
 
+/// Returns mapping of each position (`String`) to the expression (`Expr`) it is
+/// aliasing.
+pub(crate) fn extract_positions(exprs: &[Expr]) -> HashMap<String, Expr> {
+    let mut position = 0;
+    exprs
+        .iter()
+        .filter_map(|expr| match expr {
+            // position starts with 1
+            Expr::Alias(nested_expr, _alias_name) => {
+                position += 1;
+                Some((position.clone().to_string(), *nested_expr.clone()))
+            }
+            _ => {
+                position += 1;
+                Some((position.clone().to_string(), expr.clone()))
+            }
+        })
+        .collect::<HashMap<String, Expr>>()
```

Comment:
I think there might be no need to build a hashmap as we can index the expression slice directly.